### PR TITLE
[Spark] Add StringStartsWith filter pushdown to ExpressionUtils

### PR DESCRIPTION
**Pull request by viirya**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Spark's StringStartsWith(attribute, value) filter was not converted to a Kernel predicate, meaning WHERE col LIKE 'prefix%' queries had no filter pushdown in the V2 connector. The Kernel's STARTS_WITH predicate maps directly: STARTS_WITH(column, string_literal).

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by viirya** [comment]
cc @gengliangwang

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java [hunk 1]
```diff
@@ -130,6 +131,13 @@ static ConvertedPredicate convertSparkFilterToKernelPredicate(
       return new ConvertedPredicate(
           Optional.of(new Predicate("IS_NOT_NULL", kernelColumn(f.attribute()))));
     }
+    if (filter instanceof StringStartsWith) {
+      StringStartsWith f = (StringStartsWith) filter;
+      return new ConvertedPredicate(
+          Optional.of(
+              new Predicate(
+                  "STARTS_WITH", kernelColumn(f.attribute()), Literal.ofString(f.value()))));
```
Comment (gengliangwang) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java]: Nit: shall we use `convertValueToKernelLiteral` here? Also, will Spark pushdown null values?

Comment (viirya) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java]: Fixed. Using convertValueToKernelLiteral here. Also handling null value now (added a test case too).

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java [hunk 2]
```diff
@@ -130,6 +131,13 @@ static ConvertedPredicate convertSparkFilterToKernelPredicate(
       return new ConvertedPredicate(
           Optional.of(new Predicate("IS_NOT_NULL", kernelColumn(f.attribute()))));
     }
+    if (filter instanceof StringStartsWith) {
+      StringStartsWith f = (StringStartsWith) filter;
+      Optional<Literal> literal = convertValueToKernelLiteral(f.value());
+      if (!literal.isPresent()) return new ConvertedPredicate(Optional.empty());
```
Comment (gengliangwang) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java]: Nit: the other filter conversions (e.g. `EqualTo`, `GreaterThan`, etc.) use a functional style with `.map()` on the Optional:

```java
return new ConvertedPredicate(
    convertValueToKernelLiteral(f.value())
        .map(l -> new Predicate("STARTS_WITH", kernelColumn(f.attribute()), l)));
```

This would be more concise and consistent with the surrounding code.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkGoldenTableTest.java [hunk 3]
```diff
@@ -205,10 +205,15 @@ public void testDsv2Internal() throws Exception {
           new StringEndsWith("city", "York"),
         },
         // expected pushed filters
-        new Filter[] {new GreaterThan("cnt", 10), new EqualTo("date", "2025-09-01")},
+        new Filter[] {
+          new GreaterThan("cnt", 10),
+          new StringStartsWith("name", "foo"),
```
Comment (gengliangwang) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkGoldenTableTest.java]: The comment on line 197 above still says `// unsupported data filter` for `StringStartsWith("name", "foo")`, but this PR makes it a supported filter. Should be updated to `// supported data filter`.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java [hunk 4]
```diff
@@ -153,11 +153,19 @@ public void testPushFilters_singleUnsupportedDataFilter(@TempDir File tempDir) t
     checkSupportsPushDownFilters(
         builder,
         new Filter[] {new StringStartsWith("name", "test")}, // input filters
-        new Filter[] {new StringStartsWith("name", "test")}, // expected post-scan filters
-        new Filter[] {}, // expected pushed filters
-        new Predicate[] {}, // expected pushed kernel predicates
+        new Filter[] {
+          new StringStartsWith("name", "test")
+        }, // expected post-scan filters (data filter)
+        new Filter[] {new StringStartsWith("name", "test")}, // expected pushed filters
```
Comment (gengliangwang) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java]: Now that `StringStartsWith` is supported, the method name `testPushFilters_singleUnsupportedDataFilter` is misleading — the filter IS pushed down. Consider renaming it (e.g. `testPushFilters_singleStringStartsWithDataFilter`).

Same issue with:
- `testPushFilters_ORSupportedAndUnsupportedDataFilters` (line 413) — both sides of the OR are now supported.
- `testPushFilters_NOTSupportedDataORUnsupportedDataFilters` (line 666) — same.

Also, in `testPushFilters_mixedFilters`, the comments on lines 343 and 350 still say `// data filter, unsupported` for `StringStartsWith` but the pushed-filters and data-filters sections below correctly say `// data filter, supported`. These stale comments should be updated for consistency.

Comment (huan233usc) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java:177]: [AI Generated] `testPushFilters_singleUnsupportedDataFilter` now tests a supported filter (`StringStartsWith` is pushable), so we've lost test coverage for the "single unsupported data filter" scenario. The same issue repeats in several other test methods whose names reference "unsupported" but now use only supported filters:
- `testPushFilters_mixedSupportedAndUnsupportedDataFilters` (line 201)
- `testPushFilters_ORSupportedAndUnsupportedDataFilters` (line 413)
- `testPushFilters_NOTSupportedDataORUnsupportedDataFilters` (line 666)

Consider keeping the original unsupported scenarios by swapping `StringStartsWith` with `StringEndsWith` (which is genuinely unsupported), and adding separate new tests for StringStartsWith pushdown. This preserves the coverage for unsupported-filter edge cases while also verifying the new behavior.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java [hunk 5]
```diff
@@ -457,19 +500,24 @@ public void testPushFilters_mixedORandAND(@TempDir File tempDir) throws Exceptio
               new GreaterThan("dep_id", 2))
         },
         // expected post-scan filters
+        new Filter[] {},
```
Comment (huan233usc) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java:504]: [AI Generated] The block comment above this test (lines 478–487) is entirely stale — it still describes "unsupportedPartitionFilterB" and "partial pushdown of AND, will be converted to A OR C". Now that `StringStartsWith` is supported, all three filters are fully pushable and the result is `OR(AND(A, B), C)`. The inline comment at line 492 is also stale. Please update both to reflect the current behavior.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/utils/ExpressionUtilsTest.java [hunk 6]
```diff
@@ -109,6 +109,72 @@ public void testEqualNullSafeFilter() {
...
+  public void testStringStartsWithFilter() {
+    StringStartsWith filter = new StringStartsWith("name", "Al");
+    ExpressionUtils.ConvertedPredicate result =
+        ExpressionUtils.convertSparkFilterToKernelPredicate(filter);
+
+    assertTrue(result.isPresent(), "StringStartsWith filter should be converted");
+    assertFalse(result.isPartial(), "StringStartsWith filter should be fully converted");
+    assertEquals("STARTS_WITH", result.get().getName());
+    // Children: column + string literal
+    assertEquals(2, result.get().getChildren().size());
```
Comment (huan233usc) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/utils/ExpressionUtilsTest.java:127]: [AI Generated] These 5 StringStartsWith tests can be consolidated:

1. **Parameterize tests 1 & 2** (basic + empty prefix) — same structure, different prefix values:
```java
static Stream<Arguments> stringStartsWithProvider() {
    return Stream.of(
        Arguments.of("non-empty prefix", "Al"),
        Arguments.of("empty prefix", ""));
}

@ParameterizedTest(name = "StringStartsWith with {0} should be converted")
@MethodSource("stringStartsWithProvider")
public void testStringStartsWithFilter(String desc, String prefix) {
    StringStartsWith filter = new StringStartsWith("name", prefix);
    ExpressionUtils.ConvertedPredicate result =
        ExpressionUtils.convertSparkFilterToKernelPredicate(filter);
    assertTrue(result.isPresent());
    assertFalse(result.isPartial());
    assertEquals("STARTS_WITH", result.get().getName());
    assertEquals(2, result.get().getChildren().size());
}
```
2. **Keep the null test** as a simple standalone.
3. **Drop `testStringStartsWithFilter_InAndFilter` and `testStringStartsWithFilter_NotFilter`** — these duplicate existing coverage from `testAndFilter` / `testNotFilter` which already prove any convertible leaf filter works in compound contexts.

5 tests → 2, no coverage loss.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java (hunk 2 above)
Comment (viirya) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/ExpressionUtils.java]: ExpressionUtils.java now uses the .map() functional style for StringStartsWith

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java (hunk 4 above)
Comment (viirya) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java:177]: The four SparkScanBuilderTest methods that said "unsupported" but used StringStartsWith now use StringEndsWith with corrected expectations. A new testPushFilters_singleSupportedDataFilter_StringStartsWith test explicitly covers the new pushdown behavior.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java (hunk 5 above)
Comment (viirya) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkScanBuilderTest.java:504]: The block comment on testPushFilters_mixedORandAND is updated to reflect that all three filters are now fully supported.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/utils/ExpressionUtilsTest.java (hunk 6 above)
Comment (viirya) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/utils/ExpressionUtilsTest.java:127]: 5 StringStartsWith tests in ExpressionUtilsTest reduced to 2 — a @ParameterizedTest for non-empty and empty prefixes, plus the null-value standalone test.

File: spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkGoldenTableTest.java
Comment (seewishnew) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkGoldenTableTest.java]: Minor Nit: This should be marked as a supported filter now with this changeset.

Comment (viirya) [reply spark/v2/src/test/java/io/delta/spark/internal/v2/read/SparkGoldenTableTest.java]: Fixed. Thanks.

**Comment by viirya** [comment]
Thanks @gengliangwang @huan233usc @seewishnew
